### ISSUE TITLE
React 16 support

### ISSUE THREAD

**Issue by adidahiya** [palantir/blueprint#866]
Preemptively creating an umbrella issue for this task, like https://github.com/ant-design/ant-design/issues/5377.

Related React issues / deprecations:

- https://github.com/facebook/react/issues/8854
- https://github.com/facebook/react/issues/9207

We should try to support both v15 and v16 in a single Blueprint version (maintaining multiple compatible releases will be cumbersome).

Sub-tasks:

- [ ] migrate to [react-transition-group](https://github.com/reactjs/react-transition-group)
- [x] remove usage of proptypes (https://github.com/palantir/blueprint/issues/991)
- [ ] use `unstable_createPortal ` instead of `unstable_renderSubtreeIntoContainer` (#1509)

**Comment by giladgray** [comment]
i'll investigate the react 16 upgrade after #991 (updating to 15.5 and resolving deprecation warnings)

**Comment by luisatmaniak** [comment]
Will blueprint 1.x ever be compatible with React 16?

**Comment by L42y** [comment]
I guess `unstable_createPortal` should be `ReactDOM.createPortal` now: https://facebook.github.io/react/docs/portals.html

**Comment by awjreynolds** [comment]
React 16 support has been sign posted for milestone v2.0.0 

Given the 26 open issues on v2.0.0 this implies a very long wait for this. No criticism just an observation. Would it be useful to create a v2.0.0 release that only includes the fixes needed for React 16 and move the other 25 issues to v3.0.0?

I realise that is a pain, but I suspect a lot of package maintainers are going to get similar requests. Your work is awesome. Massively appreciate the quality of the framework.

**Comment by adidahiya** [comment]
@awjreynolds we'll try to make a release of blueprint that's compatible with React 16 before we do all the feature work in the 2.0 milestone. So we might make a release that supports 0.14.x, 15.x, and 16.x all at once on the 1.x track. Or it might be blueprint core 2.0 as you suggested, and the other features/breaking changes come in 3.0. We're looking into this in the next couple weeks.

**Comment by giladgray** [comment]
working on React 16 support... easy to add a `peerDependency` alongside 15 but I'm wondering how we want to go about ensuring support through automated tests? do we want to run the test suite against react 16 _also_ or _instead_ of 15 or _not at all (save for 2.0)_?

but wait, some serious gotchas:
- currently we run our unit suite against react 15.x using enzyme 2.x.
- we must [migrate](http://airbnb.io/enzyme/docs/guides/migration-from-2-to-3.html) to enzyme 3.x for react 16 support.
- enzyme 3.x was released last week, and `@types/enzyme` has not yet been updated 😢.

**Comment by adidahiya** [comment]
> ensuring support through automated tests

Yes, ideally we should run tests against both versions of React that we support. A good way to do this is with multiple containers in CircleCI. Look at the tslint repo as an example -- we run tests with multiple versions of Node and `tsc`.

> we must migrate to enzyme 3.x for react 16 support

For now, we could compile against with Enzyme 2.x types. Is there some subset of the test suite that doesn't use enzyme which we could run in the container where React 16 is installed? I'd really like to optimize for speed here and get a version of Blueprint out that at least doesn't generate peer dependency warnings with React 16, even if we have to say that official support (with Enzyme 3.x running our tests) is coming later (doesn't necessarily have to be the next major version, just whenever we finish migrating the tests).

**Comment by marekdlugos** [comment]
Sorry guys, would you be that kind to just explain a person who is just starting with React (me) to understand how to solve this issue?

I can't get rid off those 2 lines:
```
npm WARN pure-render-decorator@1.2.1 requires a peer of react@^15.0.0 || ^0.14.0 but none was installed.
npm WARN react-addons-css-transition-group@15.6.2 requires a peer of react@^15.4.2 but none was installed.
```

**Comment by adidahiya** [comment]
@marekdlugos see https://github.com/palantir/blueprint/issues/201#issuecomment-338253728. you can safely ignore those warnings

**Comment by mayaah** [comment]
Hi, I'm trying to use a MultiSelect but keep getting ` '@blueprintjs/core' does not contain an export named 'MultiSelect'` errors. I tried to update to v2 but every time I install Blueprint similar to @marekdlugos above, I get the same errors:
```
npm WARN pure-render-decorator@1.2.1 requires a peer of react@^15.0.0 || ^0.14.0 but none was installed.
npm WARN react-addons-css-transition-group@15.6.2 requires a peer of react@^15.4.2 but none was installed.
```

And I also tried the recommended commands if you see `UNMET PEER DEPENDENCY` errors too. 

Just wondering what I can do since I really want to use this MultiSelect haha. Thank you!

**Comment by giladgray** [comment]
@mayaah a few things:
1) if you're using React 16, you gotta be on [Blueprint 2.0](http://blueprintjs.com/docs/v2/) 👌 
1) `MultiSelect` lives in the 2.0 [`select`](http://blueprintjs.com/docs/v2/#select) package, not in `core` 📦  
1) `pure-render-decorator` dependency has been removed in 2.0 #1885  🙅‍♂️ 

hope that helps 😄

**Comment by adidahiya** [comment]
> if you're using React 16, you gotta be on Blueprint 2.0

this is not strictly true. react 16 works with the latest 1.x packages.

**Comment by giladgray** [comment]
@shravan2x the 1.x packages do not depend on r-t-g (they still use react-addons-css-transition-group) so if you're seeing that dependency then you must be bringing in a 2.0 package somehow.

Are you perhaps using labs@0.15? That one depends on core 2.0; downgrade to 0.14.

**Comment by shravan2x** [comment]
@giladgray I'm not using labs or any other 2.0 package that I know of. These are the ones I'm using:

```
"@blueprintjs/core": "^1.35.4",
"@fortawesome/fontawesome": "^1.1.3",
"@fortawesome/fontawesome-free-solid": "^5.0.6",
"@fortawesome/react-fontawesome": "0.0.17",
"babel-cli": "^6.26.0",
"babel-loader": "^7.1.2",
"babel-plugin-transform-runtime": "^6.23.0",
"babel-preset-env": "^1.6.1",
"babel-preset-react": "^6.24.1",
"babel-preset-stage-0": "^6.24.1",
"better-npm-run": "^0.1.0",
"classnames": "^2.2.5",
"clean-webpack-plugin": "^0.1.17",
"copy-webpack-plugin": "^4.2.0",
"css-loader": "^0.28.7",
"debug": "^3.1.0",
"extract-text-webpack-plugin": "^3.0.2",
"file-loader": "^1.1.5",
"html-webpack-plugin": "^2.30.1",
"path": "^0.12.7",
"postcss-loader": "^2.0.8",
"prop-types": "^15.6.0",
"react": "^16.2.0",
"react-addons-css-transition-group": "^15.6.2",
"react-dom": "^16.2.0",
"react-redux": "^5.0.6",
"react-transition-group": "^1.1.2",
"redux": "^3.7.2",
"style-loader": "^0.19.0",
"url-loader": "^0.6.2",
"webpack": "^3.8.1"
```

**Comment by giladgray** [comment]
Hmm then you shouldn't need that r-t-g dependency at all as you already have the add-ons one. Have you tried removing it?

**Comment by shravan2x** [comment]
This is what happens when I do:

```js
Uncaught Error: Cannot find module "react-transition-group/CSSTransitionGroup"
    at webpackMissingModule (index.js:10)
    at Object.__webpack_exports__.b (index.js:10)
    at __webpack_require__ (bootstrap 3799b691625e1c110800:19)
    at Object.<anonymous> (index.js:8)
    at __webpack_require__ (bootstrap 3799b691625e1c110800:19)
    at Object.<anonymous> (props.js:45)
    at __webpack_require__ (bootstrap 3799b691625e1c110800:19)
    at Object.__webpack_exports__.a (contextMenu.js:1)
    at __webpack_require__ (bootstrap 3799b691625e1c110800:19)
    at Object.defineProperty.value (tetherUtils.js:71)
webpackMissingModule @ index.js:10
__webpack_exports__.b @ index.js:10
__webpack_require__ @ bootstrap 3799b691625e1c110800:19
(anonymous) @ index.js:8
__webpack_require__ @ bootstrap 3799b691625e1c110800:19
(anonymous) @ props.js:45
__webpack_require__ @ bootstrap 3799b691625e1c110800:19
__webpack_exports__.a @ contextMenu.js:1
__webpack_require__ @ bootstrap 3799b691625e1c110800:19
Object.defineProperty.value @ tetherUtils.js:71
__webpack_require__ @ bootstrap 3799b691625e1c110800:19
(anonymous) @ app.3799b691625e1c110800.js:2728
__webpack_require__ @ bootstrap 3799b691625e1c110800:19
(anonymous) @ HeaderContainer.jsx:6
__webpack_require__ @ bootstrap 3799b691625e1c110800:19
Object.defineProperty.value @ index.js:1
__webpack_require__ @ bootstrap 3799b691625e1c110800:19
(anonymous) @ AppContainer.jsx:8
__webpack_require__ @ bootstrap 3799b691625e1c110800:19
Object.defineProperty.value @ index.js:1
__webpack_require__ @ bootstrap 3799b691625e1c110800:19
(anonymous) @ main.jsx:9
__webpack_require__ @ bootstrap 3799b691625e1c110800:19
(anonymous) @ app.3799b691625e1c110800.js:10173
__webpack_require__ @ bootstrap 3799b691625e1c110800:19
(anonymous) @ bootstrap 3799b691625e1c110800:62
(anonymous) @ bootstrap 3799b691625e1c110800:62
```

**Comment by adidahiya** [comment]
@shravan2x please open a new issue if you need support, this is getting out of scope.
